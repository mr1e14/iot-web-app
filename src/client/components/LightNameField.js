import React from "react";
import TextField from "@material-ui/core/TextField";
import { validateLightName } from "./deviceFunctions";
import { MAX_LIGHT_NAME_LENGTH, INPUT_DEBOUNCE_TIME } from "../config";
import debounce from "awesome-debounce-promise";

class LightNameField extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isValid: validateLightName(props.name)
    };
    this.saveName = debounce(
      name => this.props.handleChange(name),
      INPUT_DEBOUNCE_TIME
    );
  }

  onNameChange = event => {
    const newName = event.target.value;
    const isValid = validateLightName(newName);
    this.setState(
      {
        isValid
      },
      async () => {
        if (isValid) {
          await this.saveName(newName);
        }
      }
    );
  };
  render() {
    const { isValid } = this.state;
    const { name } = this.props;
    return (
      <TextField
        defaultValue={name}
        onChange={e => this.onNameChange(e)}
        error={!isValid}
        helperText={
          !isValid
            ? `Must be between 1 - ${MAX_LIGHT_NAME_LENGTH} characters`
            : ""
        }
        inputProps={{ className: this.props.classes.textCenter }}
      />
    );
  }
}

LightNameField.defaultProps = {
  name: ""
};

export default LightNameField;
