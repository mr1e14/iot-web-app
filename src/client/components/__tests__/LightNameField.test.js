import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import TextField from "@material-ui/core/TextField";

import LightNameField from "../LightNameField";

const classes = {
  textCenter: {
    textAlign: "center"
  }
};

const handleChange = jest.fn();

describe("LightNameField", () => {
  it("renders correctly", () => {
    const component = renderer.create(
      <LightNameField
        handleChange={handleChange}
        classes={classes}
        name="bedroom"
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("flags errors when invalid name is given", () => {
    const wrapper = mount(
      <LightNameField name="" classes={classes} handleChange={handleChange} />
    );
    const textField = wrapper.find(TextField);
    expect(textField.prop("error")).toBe(true);
    expect(textField.prop("helperText").length).toBeGreaterThan(0);
  });
  it("does not flag errors when valid name is given", () => {
    const wrapper = mount(
      <LightNameField
        name="bedroom"
        classes={classes}
        handleChange={handleChange}
      />
    );
    const textField = wrapper.find(TextField);
    expect(textField.prop("error")).toBe(false);
    expect(textField.prop("helperText").length).toBe(0);
  });
  it("changes isValid state based on input", () => {
    const wrapper = mount(
      <LightNameField
        name="bedroom"
        classes={classes}
        handleChange={handleChange}
      />
    );
    const textField = wrapper.find(TextField);
    textField.props().onChange({ target: { value: "" } });
    wrapper.update();
    expect(wrapper.instance().state.isValid).toEqual(false);

    textField.props().onChange({ target: { value: "hallway" } });
    wrapper.update();
    expect(wrapper.instance().state.isValid).toEqual(true);
  });
});
