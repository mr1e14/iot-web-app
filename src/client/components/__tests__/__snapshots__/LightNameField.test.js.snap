// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LightNameField renders correctly 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      className="MuiInputBase-input MuiInput-input textAlign"
      defaultValue="bedroom"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
    />
  </div>
  
</div>
`;
